### PR TITLE
[WIP] Deprecate postprocessor

### PR DESCRIPTION
**Pull request by JSS95**
#### References to other Issues or PRs

Related to #19463

#### Brief description of what is fixed or changed

Deprecate postprocessor and change `Add` and `Mul` in code to `add` and `mul`

#### Other comments

One thing should be noticed; this makes addition of non-commutative `Symbol` and matrix impossible.
In my opinion, `MatrixSymbol` must be used instead of `Symbol` in this case. non-commutative `Symbol` is scalar (its `.is_scalar` attribute gives `True`), and clearly it is wrong to allow addition of scalar and matrix.

#### Release Notes

- matrices
    - Matrices cannot be added or multiplied by `Add` and `Mul` anymore. `add`/`mul` or `MatAdd`/`MatMul` should be used instead.

### CODE REVIEW

**Comment by JSS95** [comment]
Some problems encountered: methods such as `subs` rely on postprocessor. They mask the matrix with `Dummy`, build `Add` or `Mul`, then replace the dummies back to get `MatAdd` and `MatMul` by postprocessor.

I think dedicated `MatrixDummy` class and replace/subs method are be required to remove postprocessor.

**Comment by oscarbenjamin** [comment]
> I think dedicated `MatrixDummy` class and replace/subs method are required to remove postprocessor.

Maybe but I wonder if we can eliminate the places in the codebase that depend on being able to do this. They cause problems in other situations as well.

**Comment by JSS95** [comment]
I'll try.

**Comment by JSS95** [comment]
@oscarbenjamin 
`MatrixExpr.from_index_summation` avidly use this kind of substitution.

```python
>>> from sympy import symbols, MatrixSymbol, trace, MatrixExpr
>>> from sympy.abc import a, b
>>> n = symbols('n', integer=True)
>>> A = MatrixSymbol("A", n, n)
>>> B = MatrixSymbol("B", n, n)
>>> C = MatrixSymbol("C", n, n)

>>> expr = (B*C)[a, b]*trace(A)
>>> expr
Sum(B[a, _i_1]*C[_i_1, b], (_i_1, 0, n - 1))*Trace(A)
>>> MatrixExpr.from_index_summation(expr)
Trace(A)*B*C
```

Here, `expr` is `Mul`. but `Sum([...])` is replaced to `B*C` to construct `Mul(B*C, trace(A)`, which is postprocessed to return `MatMul(trace(A), B, C)`.

I cannot come up with any nice idea to fix this...

**Comment by oscarbenjamin** [comment]
Is there not an appropriate place to use `mul` instead of `Mul`?

**Comment by JSS95** [comment]
`mul` itself does not support unevaluated object, so we need some other approach to resolve this matter.

**Comment by oscarbenjamin** [comment]
> Here, `expr` is `Mul`. but `Sum([...])` is replaced with `B*C` to construct `Mul(B*C, trace(A))`, which is postprocessed to return `MatMul(trace(A), B, C)`.

The sum represents a non-matrix and this is substituting it for a matrix which seems flakey to me.

I've been wondering about whether we should have `Mul.func == mul` to be able to support that sort of thing.

Otherwise though I think `from_index_summation` should be fixed so that it doesn't try to replace scalars with matrices.

**Comment by JSS95** [comment]
OK, I managed to fix `from_index_summation` without using replacement.

File: sympy/core/basic.py [hunk 1]
```diff
@@ -1804,11 +1799,21 @@ def _exec_constructor_postprocessors(cls, obj):
             except TypeError:
                 pass
 
-        for f in postprocessors.get(clsname, []):
+        processors = postprocessors.get(clsname, [])
+        if processors:
+            SymPyDeprecationWarning(
```
Comment (sylee957) [inline sympy/core/basic.py:1804]: This could already be a private API, so no deprecation warning may be needed

**Comment by JSS95** [comment]
I found out another error. Currently, unevaluated integration of matrix (which is used in `stats` module) is delegated to `Integral`. This needs to be fixed since `Integral` does not have any `shape` or `rank` attribute, and is considered as scalar. Scalar product of matrix integral returns `Mul` instead of `MatMul` because of this.

I am thinking of introducing `MatrixIntegral` and dispatching `integrate` in another PR. @oscarbenjamin @sylee957 , can I have your idea?

**Comment by oscarbenjamin** [comment]
The same problem will happen in lots of places e.g. Integral, Sum, Derivative, Subs, Piecewise etc. I don't think it's scalable to make a Matrix version of every class.

We need a generic way to query whether or not an object is a matrix. Probably `Integral._add_handler` should defer to the `_add_handler` of the integrand.

**Comment by JSS95** [comment]
@oscarbenjamin 
Even if we modify the handler to make `mul(matrixintegral, ...)` return `MatMul`, the shape of matrix integral will be checked in `MatMul.__new__`, which will raise error because `Integral` doesn't have this attribute. Then we'll have to start adding `rows`, `cols`, `shape`, etc to `Integral` and every other classes you mentioned. I fear that this will greatly maim the integrity of core.

**Comment by oscarbenjamin** [comment]
How about:

1. Integral defers to its integrand for `_add_handler`:
```
    @property
    def _add_handler(self):
        return self.function._add_handler
```
2. We can make a single dispatch `shape` function in the matrices module that gets the shape of a matrix and overload it for things like `Integral`.

Would that work?

I have some ideas about what is a better solution in general for this kind of situation but I haven't had time to write them down in detail...

**Comment by JSS95** [comment]
I think that will do for this PR.

**Comment by sylee957** [comment]
What kind of matrix integration problem does the stats module uses?

**Comment by JSS95** [comment]
@sylee957 It's involved in evaluating `ExpectationMatrix`. 
```python
>>> from sympy import Matrix
>>> from sympy.stats import Normal, ExpectationMatrix
>>> N11, N12 = Normal('N11', 11, 1), Normal('N12', 12, 1)
>>> N21, N22 = Normal('N21', 21, 1), Normal('N22', 22, 1)
>>> M11, M12 = Normal('M11', 1, 1), Normal('M12', 2, 1)
>>> M21, M22 = Normal('M21', 3, 1), Normal('M22', 4, 1)
>>> x1 = Matrix([[N11, N12], [N21, N22]])
>>> x2 = Matrix([[M11, M12], [M21, M22]])
>>> ExpectationMatrix(x1 + x2).doit()
```
(excerpted from docstring of `ExpectationMatrix`)

This involves `Integral(Matrix([[N11+1, N12+2], [M21+N21, M22+N22]]), M12)`, which is multiplied by some coefficient. Evaluation of this expression involves postprocessor.

**Comment by JSS95** [comment]
I tried to resolve it by monkey-patching `shape`, but having non-matrix argument in `MatMul` and making it considered as matrix raised endless errors. The load of work for this is almost tantamount to introducing matrix version of classes, which I'd rather prefer in this situation. I'll try that way and report here.

**Comment by oscarbenjamin** [comment]
I have a suggestion for this in #20158

**Comment by czgdp1807** [comment]
@JSS95 Are you planning to conclude this sometime in future? Please let us know, though it is necessary to complete this right away.

**Comment by JSS95** [comment]
@czgdp1807 #20158 needs to be done before this, which will take a long time (> 1 yr, I guess). And by then, probably we will get rid of postprocessor with very different approach. I will close this PR.

**Pull request closed without merging**